### ISSUE TITLE
Issue #1: Implement Instrumental Variables (IV) Estimator

### ISSUE THREAD

**Issue by datablogin** [datablogin/causal-inference-marketing#32]
## Summary
Implement a comprehensive Instrumental Variables (IV) estimator for handling unmeasured confounding in causal inference. This is the highest priority missing method from our analysis of the 'Causal Inference: What If' book implementations.

## Background
Instrumental Variables estimation addresses unmeasured confounding by using instruments that affect treatment but not the outcome directly. This method is critical for marketing analytics where unmeasured customer preferences or behaviors may confound treatment-outcome relationships.

### Core Implementation
- [ ] Create `IVEstimator` class inheriting from `BaseEstimator`
- [ ] Implement Two-Stage Least Squares (2SLS) algorithm
- [ ] Support for binary, continuous, and categorical treatments
- [ ] Wald estimator for simple IV cases
- [ ] Integration with existing Pydantic data models (`TreatmentData`, `OutcomeData`, `CovariateData`)

### Additional Requirements
- [ ] Add `InstrumentData` Pydantic model for instrument validation
- [ ] Weak instrument diagnostics (F-statistics, Stock-Yogo critical values)
- [ ] Local Average Treatment Effect (LATE) estimation
- [ ] Bootstrap confidence intervals
- [ ] Integration with G-estimation methods for IV
- [ ] Comprehensive error handling and edge case management

### API Design
```python
from causal_inference.estimators.iv import IVEstimator

estimator = IVEstimator(
    first_stage_model="logistic",  # or "linear", "random_forest"
    weak_instrument_threshold=10.0,
    bootstrap_samples=1000,
    random_state=42
)

# Fit the model
estimator.fit(treatment, outcome, covariates, instrument)

# Estimate effects
effect = estimator.estimate_ate()
late = estimator.estimate_late()

# Diagnostics
weak_test = estimator.weak_instrument_test()
first_stage = estimator.get_first_stage_diagnostics()
```

### Performance KPIs
1. **Accuracy**: ATE estimation within ±5% of reference implementation on NHEFS dataset
2. **First Stage F-Statistic**: F > 10 for strong instrument detection
3. **Computational Performance**: Complete estimation on NHEFS dataset (n=1629) in <10 seconds
4. **Memory Efficiency**: Peak memory usage <500MB for bootstrap with 1000 samples

### Statistical Validation KPIs
1. **Weak Instrument Detection**: Correctly identify weak instruments (F < 10) with 95% accuracy
2. **Confidence Interval Coverage**: 95% CI should contain true effect in 94-96% of simulated datasets
3. **Robustness**: Consistent results across different random seeds (coefficient of variation <2%)

### Code Quality KPIs
1. **Test Coverage**: ≥95% line coverage for IV estimator module
2. **Type Safety**: 100% mypy compliance with strict typing
3. **Documentation**: All public methods documented with examples
4. **Performance**: All tests complete in <60 seconds

### Test Case 1: Standard IV Estimation
```python
# Load NHEFS data with education as instrument for smoking cessation
loader = NHEFSDataLoader()
data = loader.load_processed_data()

# Define variables
treatment = TreatmentData(values=data['qsmk'], treatment_type="binary")
outcome = OutcomeData(values=data['wt82_71'], outcome_type="continuous") 
covariates = CovariateData(values=data[['age', 'sex', 'race']], names=['age', 'sex', 'race'])
instrument = InstrumentData(values=data['education'], instrument_type="continuous")

# Expected results (benchmarks from reference implementation)
expected_ate = 3.44  # kg weight change
expected_ci_lower = 0.58
expected_ci_upper = 6.30
expected_first_stage_f = 15.2
```

### Test Case 2: Weak Instrument Detection
```python
# Create weak instrument (random noise)
weak_instrument = InstrumentData(
    values=np.random.normal(0, 1, len(data)), 
    instrument_type="continuous"
)

# Should detect weak instrument and raise appropriate warning
estimator = IVEstimator()
estimator.fit(treatment, outcome, covariates, weak_instrument)
weak_test = estimator.weak_instrument_test()
assert weak_test['f_statistic'] < 10
assert weak_test['is_weak'] == True
```

### Test Case 3: Binary Instrument
```python
# Test with binary instrument
binary_instrument = InstrumentData(
    values=(data['education'] > data['education'].median()).astype(int),
    instrument_type="binary"
)
estimator.fit(treatment, outcome, covariates, binary_instrument)
effect = estimator.estimate_ate()
# Should produce reasonable estimate with appropriate uncertainty
```

## Implementation References
- Chapter 16 of "Causal Inference: What If" by Hernán and Robins
- jrfiedler/causal_inference_python_code repository (Chapter 16 notebook)
- linearmodels.IV2SLS for reference implementation patterns

## Definition of Done
- [ ] All unit tests pass with ≥95% coverage
- [ ] Integration tests with NHEFS dataset pass all KPI thresholds
- [ ] Type checking passes with mypy --strict
- [ ] Code formatting passes with ruff
- [ ] Documentation includes examples and mathematical background
- [ ] Performance benchmarks meet specified KPIs
- [ ] Weak instrument diagnostics validated against known test cases

## Acceptance Criteria
1. IV estimator produces ATE within 5% of reference implementation on NHEFS
2. First stage F-statistic correctly computed and matches reference
3. Weak instrument detection works with 95% accuracy on synthetic data
4. Bootstrap confidence intervals have appropriate coverage (94-96%)
5. All edge cases handled gracefully with informative error messages

## Priority: P0 (Highest)
This addresses the biggest gap in our causal inference capabilities and is foundational for handling unmeasured confounding in marketing analytics.
